### PR TITLE
033

### PR DESCRIPTION
**Pull request by tomalikem**

### CODE REVIEW

**Comment by tomalikem** [comment]
Closes #33

File: src/config.py [hunk 1]
```diff
@@ -8,10 +8,13 @@
...
-FLOYD_SOLVER_SEARCHING_TIME = 8 * 3600
-FLOYD_SOLVER_MAX_PRIORITY_MULTIPLIER = 1.5
-FLOYD_SOLVER_MAX_PATHS = 10
-FLOYD_EXTRACTOR_PERIOD_MULTIPLIER = 0.5
+FLOYD_SOLVER_SEARCHING_TIME: int = 8 * 3600
+FLOYD_SOLVER_MAX_PRIORITY_MULTIPLIER: float = 1.5
+FLOYD_SOLVER_MAX_PATHS: int = 10
+FLOYD_EXTRACTOR_PERIOD_MULTIPLIER: float = 0.5
+MAX_WALKING_TIME_IN_MINUTES: float = 10
+WALKING_SPEED = 1.0
```
Comment (tomekzaw) [inline src/config.py:16]: add `: float`

File: src/server/FlaskServer.py [hunk 2]
```diff
@@ -96,6 +80,27 @@ def handle_query_post(self, exchange, producer, request_json):
...
+            result["end_time"] = transfer["end_time"]
+            result["route_name"] = self.routes_df.at[transfer["route_id"], 'route_name']
+            result["headsign"] = self.routes_df.at[transfer["route_id"], 'headsign']
+            result["stops"] = self.get_stop_list(transfer["route_id"], transfer["start_stop_id"],
+                                                 transfer["end_stop_id"])
+        else:
+            result["type"] = "walking"
+            day_dur = datetime.strptime(transfer["end_date"], "%Y-%m-%d") - datetime.strptime(transfer["start_date"], "%Y-%m-%d")
+            hour_dur = datetime.strptime(transfer["end_time"], "%H:%M:%S") - datetime.strptime(transfer["start_time"], "%H:%M:%S")
+            result["duration_in_minutes"] = int((day_dur + hour_dur).seconds / 60)
```
Comment (tomekzaw) [inline src/server/FlaskServer.py:101]: another reason to use `datetime` instead of `date` and `time` separately

File: src/server/FlaskServer.py [hunk 3]
```diff
@@ -96,6 +80,27 @@ def handle_query_post(self, exchange, producer, request_json):
         result = connection_consumer.receive_msg()
         return result
 
+    def parse_transfers(self, transfer):
+        result = {}
+        result["start_stop"] = self.stops_df.at[transfer["start_stop_id"], 'stop_name']
```
Comment (tomekzaw) [inline src/server/FlaskServer.py:85]: inconsistent use of `"` and `'`

File: src/server/FlaskServer.py [hunk 4]
```diff
@@ -54,25 +56,7 @@ def handle_connection(self):
...
-                                                                    "start_time": t["start_time"],
-                                                                    "end_date": t["end_date"],
-                                                                    "end_time": t["end_time"],
-                                                                    "stops": self.get_stop_list(
-                                                                        t["route_id"], t["start_stop_id"],
-                                                                        t["end_stop_id"])
-                                                                }
-                                                                , transfers))
-                                      }, connection_results))
+                                      {"transfers": list(map(self.parse_transfers, transfers))}, connection_results))
```
Comment (tomekzaw) [inline src/server/FlaskServer.py:59]: list comprehension is better than `list(map(...`

File: src/examples/rmq_examples/example_connection.py [hunk 5]
```diff
@@ -34,10 +34,10 @@ def print_connections(connections):
     mocked_api = MockedApi(task_id)
     query_json = {
         "query_id": task_id,
-        "start_date": "24/5/2020",
+        "start_date": "2020-5-24",
```
Comment (tomekzaw) [inline src/examples/rmq_examples/example_connection.py:37]: should be `2020-05-24`

File: src/examples/flask_examples/example_connection.py [hunk 6]
```diff
@@ -7,10 +7,10 @@
     query_json = {
         "start_date": "2020-5-24",
```
Comment (tomekzaw) [inline src/examples/flask_examples/example_connection.py:8]: should be `2020-05-24` as well

File: src/data_provider/data/tmp/config.json [hunk 7]
```diff
@@ -1 +1 @@
-{"update_date": "2020-09-13 23:39:47"}
\ No newline at end of file
+{"update_date": "2020-10-13 00:00:04"}
```
Comment (tomekzaw) [inline src/data_provider/data/tmp/config.json]: `config.json` should be added to `.gitignore`

File: src/data_provider/Extractor.py [hunk 8]
```diff
@@ -132,3 +149,21 @@ def create_route_to_stops_dict(stop_times_df: pd.DataFrame, route_ids_df: pd.Dat
...
+        lat_1 = radians(lat_1)
+        lat_2 = radians(lat_2)
+        earth_radius = 6373000.0
+        dlon = lon_2 - lon_1
+        dlat = lat_2 - lat_1
+
+        a = sin(dlat / 2) ** 2 + cos(lat_1) * cos(lat_2) * sin(dlon / 2) ** 2
+        c = 2 * atan2(sqrt(a), sqrt(1 - a))
+
+        distance = earth_radius * c
```
Comment (tomekzaw) [inline src/data_provider/Extractor.py:166]: we could use an external dependency for such calculations, but that's OK as well

File: src/data_provider/Extractor.py [hunk 9]
```diff
@@ -20,14 +33,14 @@ def get_day_to_services_dict(calendar_df: pd.DataFrame):
         return day_to_services
 
     @staticmethod
-    def get_services_list(calendar_df):
+    def get_services_list(calendar_df) -> list:
         services = []
         for service_id, _, _, _, _, _, _, _ in calendar_df.itertuples():
             services.append(service_id)
         return services
```
Comment (tomekzaw) [inline src/data_provider/Extractor.py:40]: this really needs to be refactored, `return calendar_df.index` or `return list(calendar_df.index)` should be fine

File: src/data_provider/Extractor.py [hunk 10]
```diff
@@ -7,7 +9,18 @@ def __init__(self):
...
+            for id_2, _, lat_2, lon_2, _, _ in stops_df.itertuples():
+                duration = Extractor._get_walking_time(lon_1, lat_1, lon_2, lat_2)
+                if id_1 != id_2 and duration <= MAX_WALKING_TIME_IN_MINUTES * 60:
+                    adjacent_stops[id_1, id_2] = duration
+        return adjacent_stops
+
+
+    @staticmethod
+    def get_day_to_services_dict(calendar_df: pd.DataFrame) -> dict:
         day_to_services = {0: [], 1: [], 2: [], 3: [], 4: [], 5: [], 6: []}
```
Comment (tomekzaw) [inline src/data_provider/Extractor.py:24]:
```py
day_to_services = {i: [] for i in range(7)}
```

File: src/data_provider/Extractor.py [hunk 11]
```diff
@@ -7,7 +9,18 @@ def __init__(self):
         pass
 
     @staticmethod
-    def get_day_to_services_dict(calendar_df: pd.DataFrame):
+    def get_adjacent_stops_dict(stops_df: pd.DataFrame) -> dict:
+        adjacent_stops = {}
+        for id_1, _, lat_1, lon_1, _, _ in stops_df.itertuples():
+            for id_2, _, lat_2, lon_2, _, _ in stops_df.itertuples():
+                duration = Extractor._get_walking_time(lon_1, lat_1, lon_2, lat_2)
```
Comment (tomekzaw) [inline src/data_provider/Extractor.py:16]: this line of code runs twice for each stops pair

File: src/solver/ConnectionSolver.py [hunk 12]
```diff
@@ -121,7 +120,23 @@ def find_routes(self, path: List[int], start_time: int, start_date: date):
...
             else:
+                if (current_stop, next_stop) in self.adjacent_stops:
+                    walking_time = self.adjacent_stops[(current_stop, next_stop)]
+                    walking_df = copy(results_df)
+                    walking_df['route_id_' + str(i)] = WALKING_ROUTE_ID
+                    walking_df['index_' + str(i)] = [(
+                        WALKING_ROUTE_ID, WALKING_ROUTE_ID, WALKING_ROUTE_ID)] * len(walking_df)
+                    walking_df['departure_time_c_' + str(i)] = walking_df.apply(
+                        lambda row: row['departure_time_n_' + str(i-1)], axis=1)
+                    walking_df['departure_time_n_' + str(i)] = walking_df.apply(
```
Comment (tomekzaw) [inline src/solver/ConnectionSolver.py:138]: f-strings are the way to go here (`f'route_id_{i}'`)

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: src/data_provider/Extractor.py (hunk 11 above)
Comment (tomalikem) [reply src/data_provider/Extractor.py:16]: I know, but we neeed it for directed edges,

**Pull request merged**